OSDOCS-9882-machineconfig: Bug bacthed the Machine Config bug fixes (#77885)
openshift/openshift-docs: release_notes/ocp-4-16-release-notes.adoc

[id="ocp-4-16-machine-config-operator-bug-fixes_{context}"]
==== Machine Config Operator

> * Previously, the `MachineConfigNode` object was not created with a proper owner. As a result, the `MachineConfigNode` object could not be garbage collected, meaning that previously generated, but no longer useful, objects were not removed. With this release, the proper owner is set upon the creation of the `MachineConfigNode` Object and objects that become obsolete are available for garbage collection.(link:https://issues.redhat.com/browse/OCPBUGS-30090[*OCPBUGS-30090*])

Reviewer: ```suggestion
* Previously, the `MachineConfigNode` object was not created with a proper owner. As a result, the `MachineConfigNode` object could not be garbage collected, meaning that previously generated, but no longer useful, objects were not removed. With this release, the proper owner is set upon the creation of the `MachineConfigNode` object and objects that become obsolete are available for garbage collection.(link:https://issues.redhat.com/browse/OCPBUGS-30090[*OCPBUGS-30090*])
```

* Previously, the default value of the `nodeStatusUpdateFrequency` parameter was changed from `0s` to `10s`. This change inadvertently caused the `nodeStatusReportFrequency` to increase significantly, because the value was linked to the `nodeStatusReportFrequency` value. This resulted in high CPU usage on control plane operators and the API server. This fix manually sets the `nodeStatusReportFrequency` value to `5m`, which prevents this high CPU usage. (link:https://issues.redhat.com/browse/OCPBUGS-29713[*OCPBUGS-29713*])

> * Previously, a typo in an environment variable prevented a script from detecting if the `node.env` file was present. Because of this, the `node.env` file would be overwritten on every restart, preventing the kubelet host name from being fixed. With this fix the typo is corrected. As a result, edits to the `node.env` are now persisted across reboots. (link:https://issues.redhat.com/browse/OCPBUGS-27261[*OCPBUGS-27261*])

Reviewer: ```suggestion
* Previously, a typo in an environment variable prevented a script from detecting if the `node.env` file was present. Because of this, the `node.env` file would be overwritten on every restart, preventing the kubelet host name from being fixed. With this fix the typo is corrected. As a result, edits to the `node.env` now persist across reboots. (link:https://issues.redhat.com/browse/OCPBUGS-27261[*OCPBUGS-27261*])
```

> * Previously, when the `kube-apiserver` server Certificate Authority (CA) certificate was rotated, the MCO did not properly react and update the on-disk kubelet kubeconfig. The meant that the kubelet and some pods on the node were eventually unable to communicate with the APIserver, causing the node to enter the `NotReady` state. With this release, the Machine Config Operator (MCO) properly reacts to the change, and updates the on-disk kubeconfig such that authenticated communication with the APIServer can continue when this rotates, and also restarts kubelet/MCDaemon pod. The certificate authority has 10-year validity, so this rotation should happen rarely and is generally non-disruptive. (link:https://issues.redhat.com/browse/OCPBUGS-25821[*OCPBUGS-25821*])

Reviewer: ```suggestion
* Previously, when the `kube-apiserver` server Certificate Authority (CA) certificate was rotated, the MCO did not properly react and update the on-disk kubelet kubeconfig. This meant that the kubelet and some pods on the node were eventually unable to communicate with the APIserver, causing the node to enter the `NotReady` state. With this release, the Machine Config Operator (MCO) properly reacts to the change, and updates the on-disk kubeconfig such that authenticated communication with the APIServer can continue when this rotates, and also restarts kubelet/MCDaemon pod. The certificate authority has 10-year validity, so this rotation should happen rarely and is generally non-disruptive. (link:https://issues.redhat.com/browse/OCPBUGS-25821[*OCPBUGS-25821*])
```

* Previously, when a new node was added to or removed from a cluster, the `MachineConfigNode` (MCN) objects did not react. As a result, extraneous MCN objects existed. With this release, the Machine Config Operator removes and adds MCN objects as appropriate when nodes are added or removed. (link:https://issues.redhat.com/browse/OCPBUGS-24416[*OCPBUGS-24416*])

> * Previously, the `nodeip-configuration` service did not send logs to the serial console, which made it difficult to debug problems when networking is not available and there is no access to the node. With this release, the `nodeip-configuration` service logs output to the console for easier debugging when there is no network access to the node. (link:https://issues.redhat.com/browse/OCPBUGS-19628[*OCPBUGS-19628*])

Reviewer: presumably:
```suggestion
* Previously, the `nodeip-configuration` service did not send logs to the serial console, which made it difficult to debug problems when networking is not available and there is no access to the node. With this release, the `nodeip-configuration` service logs output to the serial console for easier debugging when there is no network access to the node. (link:https://issues.redhat.com/browse/OCPBUGS-19628[*OCPBUGS-19628*])
```
(vs. web console)

> * Previously, when a `MachineConfigPool` had the `OnBlusterBuild` functionality enabled and the `configmap` was updated with an invalid `imageBuilderType`, the machine-config ClusterOperator was not degraded. With this release, the Machine Config Operator (MCO) `ClusterOperator` status now validates the `OnBlusterBuild` inputs each time it syncs, ensuring that if those are invalid, the `ClusterOperator` is degraded. (link:https://issues.redhat.com/browse/OCPBUGS-18955[*OCPBUGS-18955*])

Reviewer: just a question: should this be `OnClusterBuild`?
```suggestion
* Previously, when a `MachineConfigPool` had the `OnClusterBuild` functionality enabled and the `configmap` was updated with an invalid `imageBuilderType`, the machine-config ClusterOperator was not degraded. With this release, the Machine Config Operator (MCO) `ClusterOperator` status now validates the `OnClusterBuild` inputs each time it syncs, ensuring that if those are invalid, the `ClusterOperator` is degraded. (link:https://issues.redhat.com/browse/OCPBUGS-18955[*OCPBUGS-18955*])
```
